### ISSUE TITLE
New Checks: "Towards a Catalog of Refactorings for Elixir"

### ISSUE THREAD

**Issue by Nezteb** [rrrene/credo-proposals#91]
### What do you want Credo to do?

See @lucasvegi's [work and accompanying research paper](https://github.com/lucasvegi/Elixir-Refactorings).

It's a carefully curated collection of 54 good Elixir refactoring opportunities, and I think many of them would make good `credo` checks.

[I mentioned this in their repo as well](https://github.com/lucasvegi/Elixir-Refactorings/issues/1).

### Which existing behaviour would change?

This would add new checks that `credo` users could opt-in to.

**Comment by lucasvegi** [comment]
I completely agree with the suggestion! We actually plan to automate checks for our catalogs of Code Smells and Refactorings at some point in the future, and Credo could be an excellent alternative for that.

**Comment by rrrene** [comment]
I am sympathetic to this as well. We could include the checks directly in Credo an give them a special tag or, if you want to remain in control of the checks' code, we could provide a plugin that includes those checks.

**Comment by lucasvegi** [comment]
> I am sympathetic to this as well. We could include the checks directly in Credo an give them a special tag or, if you want to remain in control of the checks' code, we could provide a plugin that includes those checks.

Amazing! It would be exciting to have our research incorporated into Credo. Suddenly not only the refactorings part, but also the code smells part which is being incorporated into the Elixir official documentation. As for the incorporation strategy (directly or via plugin), honestly, I still don't have an opinion on which would be best.

**Comment by Nezteb** [comment]
@rrrene Long time no chat!

I realized that I left a couple comments on another GH issue that I didn't cross-post back here.

A while back I created [an initial spreadsheet](https://docs.google.com/spreadsheets/d/1EL62b_V3aCoCo8H1BE6e4dW_tq38a0O6xJCPYnTQbj0/edit?pli=1&gid=0#gid=0) to try to map Lucas' refactorings to new Credo checks. There are still several "TBD" sections.

Another thing that would be nice to figure out: which of these checks would be easiest to implement? I ask because I'd like to take a crack at an initial MVP by implementing one or two checks. 😄

**Comment by rrrene** [comment]
Hi, still intrigued by the effort on display here. Almost 80 new checks, wow. 

I am not sure that I could pick a good starting point. I would just advise anyone undertaking this to neither pick the easiest one first nor the hardest one. I made both mistakes and they live with me for a decade now 😅

While I am much more time-constraint than two years ago, please ping me when you have a working plugin of these checks 👍

It will be great to see how you incorporated all the different/conflicting ideas from the linked repo 🙂
